lang/rust/rust.py

```diff
@@ -1,215 +1,568 @@
+conversion_traits = {
+    'into': 'Into',
+    'from': 'From',
+}
 
-    def code_operator_equal():
-        actions.auto(' == ')
+iterator_traits = {
+    'iterator': 'Iterator',
+}
 
-    def code_operator_not_equal():
-        actions.auto(' != ')
+all_traits = {
+    **closure_traits,
+    **conversion_traits,
+    **iterator_traits,
+}
 
-    def code_operator_greater_than():
-        actions.auto(' > ')
 
-    def code_operator_greater_than_or_equal_to():
-        actions.auto(' >= ')
+# tag: libraries_gui
+ctx.lists['user.code_libraries'] = {
+    'eye oh': 'std::io',
+    'file system': 'std::fs',
+    'envy': 'std::env',
+    'collections': 'std::collections',
+}
 
-    def code_operator_less_than():
-        actions.auto(' < ')
+# tag: functions_gui
+ctx.lists['user.code_functions'] = {
+    'drop': 'drop',
+    'catch unwind': 'catch_unwind',
+    'iterator': 'iter',
+    'into iterator': 'into_iter',
+    'from iterator': 'from_iter',
+    **all_macros,
+}
 
-    def code_operator_less_than_or_equal_to():
-        actions.auto(' <= ')
+# tag: functions
+ctx.lists['user.code_type'] = all_types
+
+# rust specific grammar
+ctx.lists['user.code_type_modifier'] = {
+    'mutable': 'mut ',
+    'mute': 'mut ',
+    'borrowed': '&',
+    'borrowed mutable': '&mut ',
+    'borrowed mute': '&mut ',
+    'mutable borrowed': '&mut ',
+    'mute borrowed': '&mut ',
+}
 
-    def code_operator_and():
-        actions.auto(' && ')
 
-    def code_operator_or():
-        actions.auto(' || ')
+@ctx.capture("user.code_type", rule='[{user.code_type_modifier}]{user.code_type}')
+def code_type(m) -> str:
+    """Returns a macro name"""
+    return ''.join(m)
 
-    def code_operator_bitwise_and():
-        actions.auto(' & ')
 
-    def code_operator_bitwise_or():
-        actions.auto(' | ')
+ctx.lists['user.code_macros'] = all_macros
 
-    def code_operator_bitwise_exclusive_or():
-        actions.auto(' ^ ')
+ctx.lists['user.code_trait'] = all_traits
 
-    def code_operator_bitwise_left_shift():
-        actions.auto(' << ')
 
-    def code_operator_bitwise_left_shift_assignment():
-        actions.auto(' <<= ')
+@ctx.action_class('user')
+class UserActions:
 
-    def code_operator_bitwise_right_shift():
-        actions.auto(' >> ')
+    # tag: comment_line
 
-    def code_operator_bitwise_right_shift_assignment():
-        actions.auto(' >>= ')
+    def code_comment_line_prefix():
+        actions.auto_insert('// ')
 
-    def code_operator_object_accessor():
-        actions.auto('.')
+    # tag: comment_documentation
 
-    def code_state_switch():
-        actions.insert('match ')
+    def code_comment_documentation():
+        actions.auto_insert('/// ')
+
+    # tag: imperative
 
     def code_block():
-        actions.auto('{\n\n}')
+        actions.auto_insert('{}')
         actions.edit.left()
+        actions.key('enter')
+
+    def code_state_if():
+        actions.auto_insert('if  {  }')
+        repeat_call(5, actions.edit.left)
```

Comment:
There's an `actions.user.insert_between` that was added in #781, which can be used here instead.

I'm not convinced these should insert the braces in the first place though. Having to say ~~`go right twice`~~ `go right third` (and counting to get it right in the first place) is a significant cognitive load for me, compared to saying `brace` or `block` if these didn't insert the braces.

Reply:
I didn't know about the insert between function so that is great. I've replaced use with it where appropriate. and I find that I agree with your comments on the if statement, and similar such constructs. I've removed the braces wherever a
is not inserted, such is in a for loop.